stingray/fourier.py

```diff
@@ -0,0 +1,1286 @@
+        den = (P1 - P1noise) * (P2 - P2noise)
+        num = (C * C.conj()).real - bsq
+        num[num < 0] = (C * C.conj()).real[num < 0]
+        new_coherence = num / den
+        count += 1
+
+    return new_coherence
+
+
+def error_on_averaged_cross_spectrum(C, Ps, Pr, Nave, Psnoise, Prnoise, common_ref="False"):
+    """Error on cross spectral quantities, From Ingram 2019.
+
+    Parameters
+    ----------
+    C : complex `np.array`
+        cross spectrum
+    Ps : float `np.array`
+        sub-band periodogram
+    Pr : float `np.array`
+        reference-band periodogram
+    Psnoise : float
+        Poisson noise level of the sub-band periodogram
+    Prnoise : float
+        Poisson noise level of the reference-band periodogram
+    Nave : int
+        number of intervals that have been averaged to obtain the input spectra
+
+    Other Parameters
+    ----------------
+    common_ref : bool, default False
+        Are data in the sub-band also included in the reference band?
+
+    Returns
+    -------
+    dRe : float `np.array`
+        Error on the real part of the cross spectrum
+    dIm : float `np.array`
+        Error on the imaginary part of the cross spectrum
+    dphi : float `np.array`
+        Error on the angle (or phase lag)
+    dG : float `np.array`
+        Error on the modulus of the cross spectrum
+
+    """
+    twoN = 2 * Nave
+    if common_ref:
+        Gsq = (C * C.conj()).real
+        bsq = bias_term(C, Ps, Pr, Psnoise, Prnoise, Nave)
+        frac = (Gsq - bsq) / (Pr - Prnoise)
+        PoN = Pr / twoN
+
+        # Eq. 18
+        dRe = dIm = dG = np.sqrt(PoN * (Ps - frac))
+        # Eq. 19
+        dphi = np.sqrt(PoN * (Ps / (Gsq - bsq) - 1 / (Pr - Prnoise)))
+
+    else:
+        PrPs = Pr * Ps
+        dRe = np.sqrt((PrPs + C.real ** 2 - C.imag ** 2) / twoN)
+        dIm = np.sqrt((PrPs - C.real ** 2 + C.imag ** 2) / twoN)
+        gsq = raw_coherence(C, Ps, Pr, Psnoise, Prnoise, Nave)
+        dphi = np.sqrt((1 - gsq) / (2 * gsq ** 2 * Nave))
+        dG = np.sqrt(PrPs / Nave)
+
+    return dRe, dIm, dphi, dG
+
+
+def cross_to_covariance(C, Pr, Prnoise, delta_nu):
+    """Convert a cross spectrum into a covariance.
+     Parameters
+    ----------
+    C : complex `np.array`
+        cross spectrum
+    Pr : float `np.array`
+        reference-band periodogram
+    Prnoise : float
+        Poisson noise level of the reference-band periodogram
+    delta_nu : float or `np.array`
+        spectral resolution. Can be a float, or an array if the spectral
+        resolution is not constant throughout the periodograms
+
+    """
+    return C * np.sqrt(delta_nu / (Pr - Prnoise))
+
+
+def _which_segment_idx_fun(counts=None, dt=None):
+    # Make function interface equal (counts gets ignored)
+    if counts is None:
+        return generate_indices_of_segment_boundaries_unbinned
+
+    def fun(*args, **kwargs):
+        return generate_indices_of_segment_boundaries_binned(*args, dt=dt, **kwargs)
```

Comment:
I don't understand the logic here. If `counts is None`, then it returns the bare function definition. If `counts` is *not* `None`, it'll generate a function that makes use of `dt`, but how do you know that `dt` isn't `None`? Especially since `counts` isn't actually used in `fun`?

Reply:
Ok, let's try now